File: frontend/src/types/movement.ts
export type MovementType = 'warehouse_transfer' | 'customer_deployment' | 'customer_return' | 'internal_relocation'
export type MovementStatus = 'draft' | 'preparing' | 'in_transit' | 'delivered' | 'completed' | 'cancelled'
export type MovementPriority = 'low' | 'normal' | 'high' | 'urgent'

export interface ForkliftBrief {
  id: number
  serial_number: string
  name_en: string
  status: string
  current_hour_meter: number
}

export interface CustomerBrief {
  id: number
  first_name: string
  last_name: string
  company: string | null
}

export interface UserBrief {
  id: number
  username: string
  full_name: string | null
}

export interface ContractBrief {
  id: number
  contract_number: string
  status: string
}

export interface MovementCheckpoint {
  id: number
  movement_id: number
  from_status: string | null
  to_status: string
  location: string | null
  latitude: number | null
  longitude: number | null
  notes: string | null
  user: UserBrief | null
  recorded_at: string
}

export interface Movement {
  id: number
  movement_number: string
  movement_type: MovementType
  status: MovementStatus
  priority: MovementPriority
  forklift: ForkliftBrief
  customer: CustomerBrief | null
  rental_contract: ContractBrief | null
  assigned_driver: UserBrief | null
  from_location: string
  to_location: string
  scheduled_date: string
  actual_departure: string | null
  actual_arrival: string | null
  tracking_code: string | null
  is_active: boolean
  created_at: string
  updated_at: string | null
}

export interface MovementDetail extends Movement {
  from_address: string | null
  to_address: string | null
  departure_hour_meter: number | null
  arrival_hour_meter: number | null
  notes: string | null
  internal_notes: string | null
  cancellation_reason: string | null
  created_by: number | null
  updated_by: number | null
  checkpoints: MovementCheckpoint[]
}

export interface MovementListResponse {
  items: Movement[]
  total: number
  page: number
  page_size: number
  pages: number
}

export interface MovementCreate {
  movement_type: MovementType
  forklift_id: number
  customer_id?: number | null
  rental_contract_id?: number | null
  from_location: string
  from_address?: string
  to_location: string
  to_address?: string
  scheduled_date: string
  priority?: MovementPriority
  assigned_driver_id?: number | null
  tracking_code?: string | null
  notes?: string
  internal_notes?: string
}

export interface MovementListParams {
  q?: string
  movement_type?: MovementType
  status?: MovementStatus
  forklift_id?: number
  customer_id?: number
  assigned_driver_id?: number
  priority?: MovementPriority
  is_active?: boolean
  scheduled_from?: string
  scheduled_to?: string
  page?: number
  page_size?: number
  sort?: 'created_at' | 'scheduled_date' | 'movement_number' | 'status' | 'priority'
  order?: 'asc' | 'desc'
}

File: backend/app/schemas/movement.py
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.asset_movement import MovementPriority, MovementStatus, MovementType


class CustomerBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    first_name: str
    last_name: str
    company: str | None = None


class ForkliftBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    serial_number: str
    name_en: str
    status: str
    current_hour_meter: float


class UserBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


class ContractBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    contract_number: str
    status: str


# ── Create / Update ─────────────────────────────────────────────────────────

class MovementCreate(BaseModel):
    movement_type: MovementType
    forklift_id: int
    customer_id: int | None = None
    rental_contract_id: int | None = None
    from_location: str = Field(..., min_length=1, max_length=200)
    from_address: str | None = None
    to_location: str = Field(..., min_length=1, max_length=200)
    to_address: str | None = None
    scheduled_date: date
    priority: MovementPriority = MovementPriority.NORMAL
    assigned_driver_id: int | None = None
    tracking_code: str | None = Field(default=None, max_length=100)
    notes: str | None = None
    internal_notes: str | None = None


class MovementUpdate(BaseModel):
    from_location: str | None = Field(default=None, min_length=1, max_length=200)
    from_address: str | None = None
    to_location: str | None = Field(default=None, min_length=1, max_length=200)
    to_address: str | None = None
    scheduled_date: date | None = None
    priority: MovementPriority | None = None
    assigned_driver_id: int | None = None
    tracking_code: str | None = None
    notes: str | None = None
    internal_notes: str | None = None


# ── Workflow actions ─────────────────────────────────────────────────────────

class PrepareAction(BaseModel):
    notes: str | None = Field(default=None, max_length=500)


class DepartAction(BaseModel):
    departure_hour_meter: float = Field(..., ge=0)
    notes: str | None = Field(default=None, max_length=500)


class ArriveAction(BaseModel):
    arrival_hour_meter: float = Field(..., ge=0)
    notes: str | None = Field(default=None, max_length=500)


class CompleteAction(BaseModel):
    notes: str | None = Field(default=None, max_length=500)


class CancelAction(BaseModel):
    cancellation_reason: str = Field(..., min_length=1)


class CheckpointAction(BaseModel):
    location: str | None = Field(default=None, max_length=200)
    latitude: float | None = None
    longitude: float | None = None
    notes: str | None = Field(default=None, max_length=500)


# ── Output schemas ──────────────────────────────────────────────────────────

class MovementHistoryOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    movement_id: int
    from_status: str | None = None
    to_status: str
    location: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    notes: str | None = None
    user: UserBrief | None = None
    recorded_at: datetime


class MovementOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    movement_number: str
    movement_type: str
    status: str
    priority: str
    forklift: ForkliftBrief
    customer: CustomerBrief | None = None
    rental_contract: ContractBrief | None = None
    assigned_driver: UserBrief | None = None
    from_location: str
    to_location: str
    scheduled_date: date
    actual_departure: datetime | None = None
    actual_arrival: datetime | None = None
    tracking_code: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime | None = None


class MovementDetail(MovementOut):
    from_address: str | None = None
    to_address: str | None = None
    departure_hour_meter: float | None = None
    arrival_hour_meter: float | None = None
    notes: str | None = None
    internal_notes: str | None = None
    cancellation_reason: str | None = None
    created_by: int | None = None
    updated_by: int | None = None
    checkpoints: list[MovementHistoryOut] = []


class MovementListResponse(BaseModel):
    items: list[MovementOut]
    total: int
    page: int
    page_size: int
    pages: int


MovementDetail.model_rebuild()
